File: flutter_inappwebview_platform_interface/lib/src/types/save_as_ui_showing_request.dart
import 'package:flutter_inappwebview_internal_annotations/flutter_inappwebview_internal_annotations.dart';

import 'save_as_kind.dart';
import 'enum_method.dart';

part 'save_as_ui_showing_request.g.dart';

///Class that represents the request used by the [PlatformWebViewCreationParams.onSaveAsUIShowing] event.
@ExchangeableObject()
class SaveAsUIShowingRequest_ {
  ///The MIME type of content to be saved.
  String? contentMimeType;

  ///Whether to cancel the Save As action.
  bool? cancel;

  ///Whether to suppress the default dialog.
  bool? suppressDefaultDialog;

  ///The suggested save file path.
  String? saveAsFilePath;

  ///Whether the save operation may replace an existing file.
  bool? allowReplace;

  ///The save-as kind for the operation.
  SaveAsKind_? kind;

  SaveAsUIShowingRequest_({
    this.contentMimeType,
    this.cancel,
    this.suppressDefaultDialog,
    this.saveAsFilePath,
    this.allowReplace,
    this.kind,
  });
}

File: flutter_inappwebview_windows/windows/types/save_as_ui_showing_request.cpp
#include "save_as_ui_showing_request.h"

namespace flutter_inappwebview_plugin
{
  SaveAsUIShowingRequest::SaveAsUIShowingRequest(const std::optional<std::string>& contentMimeType,
    const std::optional<bool>& cancel,
    const std::optional<bool>& suppressDefaultDialog,
    const std::optional<std::string>& saveAsFilePath,
    const std::optional<bool>& allowReplace,
    const std::optional<SaveAsKind>& kind)
    : contentMimeType(contentMimeType), cancel(cancel), suppressDefaultDialog(suppressDefaultDialog),
    saveAsFilePath(saveAsFilePath), allowReplace(allowReplace), kind(kind)
  {}

  SaveAsUIShowingRequest::SaveAsUIShowingRequest(const flutter::EncodableMap& map)
    : SaveAsUIShowingRequest(get_optional_fl_map_value<std::string>(map, "contentMimeType"),
      get_optional_fl_map_value<bool>(map, "cancel"),
      get_optional_fl_map_value<bool>(map, "suppressDefaultDialog"),
      get_optional_fl_map_value<std::string>(map, "saveAsFilePath"),
      get_optional_fl_map_value<bool>(map, "allowReplace"),
      SaveAsKindFromOptionalInteger(get_optional_fl_map_value<int64_t>(map, "kind")))
  {}

  flutter::EncodableMap SaveAsUIShowingRequest::toEncodableMap() const
  {
    return flutter::EncodableMap{
      {"contentMimeType", make_fl_value(contentMimeType)},
      {"cancel", make_fl_value(cancel)},
      {"suppressDefaultDialog", make_fl_value(suppressDefaultDialog)},
      {"saveAsFilePath", make_fl_value(saveAsFilePath)},
      {"allowReplace", make_fl_value(allowReplace)},
      {"kind", make_fl_value(SaveAsKindToInteger(kind))}
    };
  }
}